File: tools/shell/src/controller/collectionController.py
import json

from src.connection import connection
from src.environmentVariable import EnvironmentVariable
from src.constant import Constant


class CollectionController:

    __SERVICE_NAME = "collection"

    @staticmethod
    def createCollection():
        collectionName = input(
            f"{Constant.COMMAND_QUESTION_PREFIX}Collection Name: ")

        print(connection.send(CollectionController.__SERVICE_NAME, "createCollection", {
              "databaseName": EnvironmentVariable.databaseName, "collectionName": collectionName}))

    @staticmethod
    def setCollectionName():
        oldCollectionName = input(
            f"{Constant.COMMAND_QUESTION_PREFIX}Old Collection Name: ")

        newCollectionName = input(
            f"{Constant.COMMAND_QUESTION_PREFIX}New Collection Name: ")

        print(connection.send(CollectionController.__SERVICE_NAME, "setCollectionName", {
              "databaseName": EnvironmentVariable.databaseName, "collectionName": oldCollectionName, "newCollectionName": newCollectionName}))

    @staticmethod
    def getAllObjects():
        collectionName = input(
            f"{Constant.COMMAND_QUESTION_PREFIX}Collection Name: ")

        print(connection.send(CollectionController.__SERVICE_NAME, "getAllObjects", {
              "databaseName": EnvironmentVariable.databaseName, "collectionName": collectionName}))

    @staticmethod
    def findFromObjectId():
        collectionName = input(
            f"{Constant.COMMAND_QUESTION_PREFIX}Collection Name: ")

        objectId = input(f"{Constant.COMMAND_QUESTION_PREFIX}Object ID: ")

        print(connection.send(CollectionController.__SERVICE_NAME, "findFromObjectId", {
              "databaseName": EnvironmentVariable.databaseName, "collectionName": collectionName, "objectId": objectId}))
    
    @staticmethod
    def findOne():
        collectionName = input(
            f"{Constant.COMMAND_QUESTION_PREFIX}Collection Name: ")

        query = json.loads(input(Constant.COMMAND_QUESTION_PREFIX +
                                 "Query (JSON Format like {'db': 'sample_db'}): "))

        print(connection.send(CollectionController.__SERVICE_NAME, "findOne", {
              "databaseName": EnvironmentVariable.databaseName, "collectionName": collectionName, "query": query}))

    @staticmethod
    def find():
        collectionName = input(
            f"{Constant.COMMAND_QUESTION_PREFIX}Collection Name: ")

        query = json.loads(input(Constant.COMMAND_QUESTION_PREFIX +
                                 "Query (JSON Format like {'db': 'sample_db'}): "))

        print(connection.send(CollectionController.__SERVICE_NAME, "find", {
              "databaseName": EnvironmentVariable.databaseName, "collectionName": collectionName, "query": query}))

    @staticmethod
    def insert():
        collectionName = input(
            f"{Constant.COMMAND_QUESTION_PREFIX}Collection Name: ")

        data = json.loads(input(Constant.COMMAND_QUESTION_PREFIX +
                                "Data (JSON Format like {'db': 'sample_db'}): "))

        print(connection.send(CollectionController.__SERVICE_NAME, "insert", {
              "databaseName": EnvironmentVariable.databaseName, "collectionName": collectionName, "data": data}))

    @staticmethod
    def updateFromObjectId():
        collectionName = input(
            f"{Constant.COMMAND_QUESTION_PREFIX}Collection Name: ")

        objectId = input(f"{Constant.COMMAND_QUESTION_PREFIX}Object ID: ")

        updateData = json.loads(input(Constant.COMMAND_QUESTION_PREFIX +
                                      "Update Data (JSON Format like {'db': 'sample_db'}): "))

        print(connection.send(CollectionController.__SERVICE_NAME, "updateFromObjectId", {
              "databaseName": EnvironmentVariable.databaseName, "collectionName": collectionName, "objectId": objectId, "updateData": updateData}))

    @staticmethod
    def update():
        collectionName = input(
            f"{Constant.COMMAND_QUESTION_PREFIX}Collection Name: ")

        query = json.loads(input(Constant.COMMAND_QUESTION_PREFIX +
                                 "Query (JSON Format like {'db': 'sample_db'}): "))

        updateData = json.loads(input(Constant.COMMAND_QUESTION_PREFIX +
                                      "Update Data (JSON Format like {'db': 'sample_db'}): "))

        print(connection.send(CollectionController.__SERVICE_NAME, "update", {
              "databaseName": EnvironmentVariable.databaseName, "collectionName": collectionName, "query": query, "updateData": updateData}))

    @staticmethod
    def deleteFromObjectId():
        collectionName = input(
            f"{Constant.COMMAND_QUESTION_PREFIX}Collection Name: ")

        objectId = input(f"{Constant.COMMAND_QUESTION_PREFIX}Object ID: ")

        print(connection.send(CollectionController.__SERVICE_NAME, "deleteFromObjectId", {
              "databaseName": EnvironmentVariable.databaseName, "collectionName": collectionName, "objectId": objectId}))

    @staticmethod
    def delete():
        collectionName = input(
            f"{Constant.COMMAND_QUESTION_PREFIX}Collection Name: ")

        query = json.loads(input(Constant.COMMAND_QUESTION_PREFIX +
                                 "Query (JSON Format like {'db': 'sample_db'}): "))

        print(connection.send(CollectionController.__SERVICE_NAME, "delete", {
              "databaseName": EnvironmentVariable.databaseName, "collectionName": collectionName, "query": query}))

    @staticmethod
    def drop():
        collectionName = input(
            f"{Constant.COMMAND_QUESTION_PREFIX}Collection Name: ")

        print(connection.send(CollectionController.__SERVICE_NAME,
              "drop", {"databaseName": EnvironmentVariable.databaseName, "collectionName": collectionName}))

File: tools/shell/src/main/java/com/padr/tuxdb/shell/controller/CollectionController.java
package com.padr.tuxdb.shell.controller;

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.HashMap;
import java.util.Map;

import com.google.gson.Gson;
import com.padr.tuxdb.shell.Connection;
import com.padr.tuxdb.shell.Constant;
import com.padr.tuxdb.shell.EnviormentVariable;

public class CollectionController {

    private static final String SERVICE = "collection";
    private static final BufferedReader input = new BufferedReader(new InputStreamReader(System.in));

    public static void getName() throws Exception {
        System.out.println(EnviormentVariable.collectionName);
    }

    @SuppressWarnings("unchecked")
    public static void setName() throws Exception {
        System.out.print(String.format("%s New Collection Name: ", Constant.COMMAND_ANSWER_PREFIX));
        String newCollectionName = input.readLine();

        Map<String, Object> additionalParameters = new HashMap<>();
        additionalParameters.put("newCollectionName", newCollectionName);

        Map<String, Object> response = new HashMap<>();

        response = new Gson().fromJson(Connection.send(SERVICE, "setName", requestScheme(additionalParameters)), response.getClass());

        if ((boolean) (response.get("success")))
            EnviormentVariable.collectionName = newCollectionName; 

        System.out.println(response);
    }

    public static void getAllObjects() throws Exception {
        System.out.println(Connection.send(SERVICE, "getAllObjects", requestScheme(null)));
    }

    public static void findFromObjectId() throws Exception {
        System.out.print(String.format("%s Object ID: ", Constant.COMMAND_ANSWER_PREFIX));
        String objectId = input.readLine();

        Map<String, Object> additionalParameters = new HashMap<>();
        additionalParameters.put("objectId", objectId);

        System.out.println(Connection.send(SERVICE, "findFromObjectId", requestScheme(additionalParameters)));
    }

    @SuppressWarnings("unchecked")
    public static void findOne() throws Exception {
        Map<String, Object> query = new HashMap<>();

        System.out.print(String.format("%s Query (JSON Format like {'key': 'value'}): ", Constant.COMMAND_ANSWER_PREFIX));
        query = new Gson().fromJson(input.readLine(), query.getClass());

        Map<String, Object> additionalParameters = new HashMap<>();
        additionalParameters.put("query", query);

        System.out.println(Connection.send(SERVICE, "findOne",  requestScheme(additionalParameters)));
    }

    @SuppressWarnings("unchecked")
    public static void find() throws Exception {
        Map<String, Object> query = new HashMap<>();

        System.out.print(String.format("%s Query (JSON Format like {'key': 'value'}): ", Constant.COMMAND_ANSWER_PREFIX));
        query = new Gson().fromJson(input.readLine(), query.getClass());

        Map<String, Object> additionalParameters = new HashMap<>();
        additionalParameters.put("query", query);

        System.out.println(Connection.send(SERVICE, "find",  requestScheme(additionalParameters)));
    }

    @SuppressWarnings("unchecked")
    public static void insert() throws Exception {
        Map<String, Object> data = new HashMap<>();

        System.out.print(String.format("%s Data (JSON Format like {'key': 'value'}): ", Constant.COMMAND_ANSWER_PREFIX));
        data = new Gson().fromJson(input.readLine(), data.getClass());

        Map<String, Object> additionalParameters = new HashMap<>();
        additionalParameters.put("data", data);

        System.out.println(Connection.send(SERVICE, "insert",  requestScheme(additionalParameters)));
    }

    @SuppressWarnings("unchecked")
    public static void updateFromObjectId() throws Exception {
        System.out.print(String.format("%s Object ID: ", Constant.COMMAND_ANSWER_PREFIX));
        String objectId = input.readLine();

        Map<String, Object> updateData = new HashMap<>();

        System.out.print(String.format("%s Update Data (JSON Format like {'key': 'value'}): ", Constant.COMMAND_ANSWER_PREFIX));
        updateData = new Gson().fromJson(input.readLine(), updateData.getClass());

        Map<String, Object> additionalParameters = new HashMap<>();
        additionalParameters.put("objectId", objectId);
        additionalParameters.put("updateData", updateData);

        System.out.println(Connection.send(SERVICE, "updateFromObjectId",  requestScheme(additionalParameters)));
    }

    @SuppressWarnings("unchecked")
    public static void update() throws Exception {
        Map<String, Object> query = new HashMap<>();

        System.out.print(String.format("%s Query (JSON Format like {'key': 'value'}): ", Constant.COMMAND_ANSWER_PREFIX));
        query = new Gson().fromJson(input.readLine(), query.getClass());

        Map<String, Object> updateData = new HashMap<>();

        System.out.print(String.format("%s Update Data (JSON Format like {'key': 'value'}): ", Constant.COMMAND_ANSWER_PREFIX));
        updateData = new Gson().fromJson(input.readLine(), updateData.getClass());

        Map<String, Object> additionalParameters = new HashMap<>();
        additionalParameters.put("query", query);
        additionalParameters.put("updateData", updateData);

        System.out.println(Connection.send(SERVICE, "update",  requestScheme(additionalParameters)));
    }

    public static void deleteFromObjectId() throws Exception {
        System.out.print(String.format("%s Object ID: ", Constant.COMMAND_ANSWER_PREFIX));
        String objectId = input.readLine();

        Map<String, Object> additionalParameters = new HashMap<>();
        additionalParameters.put("objectId", objectId);

        System.out.println(Connection.send(SERVICE, "deleteFromObjectId",  requestScheme(additionalParameters)));
    }

    @SuppressWarnings("unchecked")
    public static void delete() throws Exception {
        Map<String, Object> query = new HashMap<>();

        System.out.print(String.format("%s Query (JSON Format like {'key': 'value'}): ", Constant.COMMAND_ANSWER_PREFIX));
        query = new Gson().fromJson(input.readLine(), query.getClass());

        Map<String, Object> additionalParameters = new HashMap<>();
        additionalParameters.put("query", query);

        System.out.println(Connection.send(SERVICE, "delete",  requestScheme(additionalParameters)));
    }

    public static void drop() throws Exception {
        System.out.println(Connection.send(SERVICE, "drop", requestScheme(null)));
    }

    private static Map<String, Object> requestScheme(Map<String, Object> additionalParameters) {
        Map<String, Object> scheme = new HashMap<>();

        scheme.put("databaseName", EnviormentVariable.databaseName);
        scheme.put("collectionName", EnviormentVariable.collectionName);

        if (additionalParameters != null)
            scheme.putAll(additionalParameters);

        return scheme;
    }
}
